src/lein_monolith/task/fingerprint.clj

```diff
@@ -0,0 +1,352 @@
+(ns lein-monolith.task.fingerprint
+  (:require
+    [clojure.data]
+    [clojure.edn :as edn]
+    [clojure.java.io :as jio]
+    [clojure.set :as set]
+    [clojure.string :as str]
+    [leiningen.core.main :as lein]
+    [leiningen.core.project :as project]
+    [lein-monolith.dependency :as dep]
+    [lein-monolith.target :as target]
+    [lein-monolith.task.util :as u]
+    [multihash.core :as mhash]
+    [multihash.digest :as digest]
+    [puget.color.ansi :as ansi]
+    [puget.printer :as puget])
+  (:import
+    (java.io
+      File
+      PushbackInputStream)))
+
+
+;; ## Options
+
+(def selection-opts
+  (assoc target/selection-opts :upstream 0 :downstream 0))
+
+
+;; ## Hashing projects' inputs
+
+(def ^:private ->multihash
+  "Globally storing the algorithm we use to generate each multihash."
+  digest/sha1)
+
+
+(defn- aggregate-hashes
+  "Takes a collection of multihashes, and aggregates them together into a unified hash."
+  [mhashes]
+  ;; TODO: is there a better way to do this?
+  (if (= 1 (count mhashes))
+    (first mhashes)
+    (->> mhashes
+         (map mhash/base58)
+         (sort)
+         (apply str)
+         (->multihash))))
+
+
+(defn- list-all-files
+  [^File file]
+  (if (.isFile file)
+    [file]
+    (mapcat list-all-files (.listFiles file))))
+
+
+(defn- hash-file
+  "Takes a File object, and returns a multihash that uniquely identifies the
+  content of this file and the location of the file."
+  [^File file]
+  ;; TODO: only prefix the file location relative to the subproject root?
+  (let [prefix (.getBytes (str (.getAbsolutePath file) "\n"))]
+    (with-open [in (PushbackInputStream. (jio/input-stream file) (count prefix))]
+      (.unread in prefix)
```

Comment:
Ah, I see how you're incorporating the paths. I'd suggest instead that you build the paths into the aggregate, not into the individual file hashes like this. Otherwise this is going to result in a new hash for every file if the user moves the subproject or the top-level monolith directory.